fix fetchFromGitLab to use repository archive downloads

justinwoo opened the pull request:
* aligns with fetchFromGitHub's use of repository archive links

* fixes broken archive fetching due to GitLab API issues

Avoids GitLab API issues such as
<https://gitlab.com/gitlab-org/gitlab-ce/issues/38537> (see also
<https://gitlab.com/gitlab-org/gitlab-ce/issues/47336#note_103430000>,
<https://gitlab.com/gitlab-com/support-forum/issues/4015>, et al)

* fixes #48215

* Allows many emacsMelpa packages to be built (e.g. shrink-path)

###### Motivation for this change

Derivations such as shrink-path.el from emacsMelpa will not build without addressing being unable to download archives from GitLab. The method of using the GitLab API which does not correctly handle repository names simply fails, whereas use of the normal GitLab repository archives works as intended and matches the usage of the GitHub archives.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

marsam commented:
IIUC we use the gitlab api because it has guaranteed stability https://github.com/NixOS/nixpkgs/pull/39467.
maybe for those broken melpa packages we can override their source with https://github.com/emacsmirror/

justinwoo commented:
This bug does mean that not only would I have to override melpa packages, but any and all GitLab packages where I run into issues like this one. I don't see the usefulness of using a "guaranteed stable" API where bugs are never fixed.

Ericson2314 commented:
Is there any way we could use the `mirror://` or whatever it is to try both? Git Lab is notoriously buggy, and both have been broken at times.

Mic92 commented:
Yeah the mirror solution sound good to me.

justinwoo committed:
fix fetchFromGitLab to use repository archive downloads

* aligns with fetchFromGitHub's use of repository archive links

* fixes broken archive fetching due to GitLab API issues

Avoids GitLab API issues such as
<https://gitlab.com/gitlab-org/gitlab-ce/issues/38537> (see also
<https://gitlab.com/gitlab-org/gitlab-ce/issues/47336#note_103430000>,
<https://gitlab.com/gitlab-com/support-forum/issues/4015>, et al)

* fixes #48215

* Allows many emacsMelpa packages to be built (e.g. shrink-path)

justinwoo commented:
Sorry, I have found that I will not be working on this PR any further. Please feel free to use the commits elsewhere if there is interest in enacting some change based off of this. Thanks.
